### PR TITLE
MODLOGSAML-63 Implement CSRF Prevention

### PR DESCRIPTION
**Pull request by hjiebsco**
Please https://issues.folio.org/browse/MODLOGSAML-63. The basic idea is to generate a random id and store in the cookie. Later then compare the value during /callback call. There is a detailed explanation in @craigmcnally's original [PR](https://github.com/folio-org/mod-login-saml/pull/63). I simplified the implementation. Hopefully it did not lower the security requirement.

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/SamlAPI.java [hunk 1]
```diff
@@ -98,11 +113,19 @@ public void getSamlCheck(RoutingContext routingContext, Map<String, String> okap
...
 
+    String csrfToken = UUID.randomUUID().toString();
+    Cookie csrfCookie = Cookie.cookie(CSRF_TOKEN, csrfToken)
+      .setPath("/").setHttpOnly(true).setSecure(true);
+    routingContext.addCookie(csrfCookie);
+
     String stripesUrl = requestEntity.getStripesUrl();
+    StringBuilder relayInfo = new StringBuilder(stripesUrl);
+    relayInfo.append(stripesUrl.indexOf('?') >= 0 ? '&' : '?')
+      .append(CSRF_TOKEN).append("=").append(csrfToken);
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/SamlAPI.java:124]: Using + is much easier to read.
The compiler automatically uses a StringBuilder internally when we use + in the code.
`String relayInfo = stripesUrl + (stripesUrl.indexOf('?') >= 0 ? '&' : '?') + CSRF_TOKEN + '=' + csrfToken;`

File: src/main/java/org/folio/rest/impl/SamlAPI.java [hunk 2]
```diff
@@ -579,4 +609,48 @@ private SamlConfig configToDto(SamlConfiguration config) {
...
+    response.putHeader(ACCESS_CONTROL_ALLOW_METHODS, request.getHeader(ACCESS_CONTROL_REQUEST_METHOD));
+    Utils.appendToMapIfAbsent(response.headers(), VARY, ",", ACCESS_CONTROL_REQUEST_METHOD);
+    response.putHeader(ACCESS_CONTROL_ALLOW_HEADERS, request.getHeader(ACCESS_CONTROL_REQUEST_HEADERS));
+    Utils.appendToMapIfAbsent(response.headers(), VARY, ",", ACCESS_CONTROL_REQUEST_HEADERS);
+    response.setStatusCode(204).end();
+  }
+
+  private void addCredentialsAndOriginHeaders(RoutingContext routingContext) {
+    String origin = routingContext.request().headers().get(ORIGIN);
+    if (origin == null || origin.isBlank() || origin.trim().contentEquals("*")) {
```
Comment (julianladisch) [inline src/main/java/org/folio/rest/impl/SamlAPI.java:648]: line 648 is the same as line 632+628 and might be extracted into `isInvalidOrigin(String)`.
line 632+628 can be combined, null is also invalid.

File: src/main/java/org/folio/rest/impl/SamlAPI.java (hunk 1 above)
Comment (hjiebsco) [reply src/main/java/org/folio/rest/impl/SamlAPI.java]: > Using + is much easier to read.
> The compiler automatically uses a StringBuilder internally when we use + in the code.
> `String relayInfo = stripesUrl + (stripesUrl.indexOf('?') >= 0 ? '&' : '?') + CSRF_TOKEN + '=' + csrfToken;`

Done.

File: src/main/java/org/folio/rest/impl/SamlAPI.java (hunk 2 above)
Comment (hjiebsco) [reply src/main/java/org/folio/rest/impl/SamlAPI.java]: > line 648 is the same as line 632+628 and might be extracted into `isInvalidOrigin(String)`.
> line 632+628 can be combined, null is also invalid.

Done.

**Review by adamdickmeiss — APPROVED** [review]

**Pull request merged**